Change event to document.body for dynamically added elements

Lenny4 opened the pull request:
Change event to document.body to allow third party plugin to add, remove, update elements. Because this approach works for dynamically added elements.

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

These changes allow a third party plugin to modify the content of the order details in javascript without having to "recall" all the javascript event linked to these elements.

Even in the future if WooCommerce starts to modify the content of `#woocommerce-order-data` this will be needed.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

1. Before checking out this branch, verify the current behavior in trunk.
1. Install this snippet in your test site: https://gist.github.com/Lenny4/448cc3894b940d3877f02754a9893cdf
2. In WP Admin, navigate to the screen to add a new order. You'll see two buttons at the top of the screen added by the snippet.
3. Click the pencil icon to edit the Billing Address for the order. It should expand to a set of fields. Then click the first "Run Test" button at the top of the screen. This will regenerate the Billing Address section, so the fields are collapsed again. Now if you click the pencil icon again, nothing will happen, because the set of fields is no longer initialized.
4. Try the same with the second "Run Test" button, and you will get the same result.
5. Now check out this branch. You may need to rebuild the admin scripts with `pnpm --filter='@woocommerce/plugin-woocommerce' build`
6. Do the same procedure with the pencil icon and the two "Run Test" buttons. The first button should still result in an unresponsive pencil icon. But the second button should result in being able to re-expand the set of billing address fields.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

Make the scripts that initialize the meta boxes on the Edit Order screen accessible to third parties

</details>

Lenny4 committed:
Change event to document.body to allow third party plugin to add, remove, update elements. Because this approach works for dynamically added elements.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Lenny4 commented:
Hi @coreymckrill ,

Any feedback about this Pull Request ?

Trank you

Lenny4 commented:
@coreymckrill up

coreymckrill commented:
Hi @Lenny4, sorry for the slow response on this one, been AFK. Could you provide some more detail in the testing instructions section? Some step-by-step instructions with which screens to load and what to click on would be helpful for me and for anyone else who tests this before it lands in a release.

> These changes allow a third party plugin to modify the content of the order details in javascript without having to "recall" all the javascript event linked to these elements.

Could you give an example of the kind of modification you're talking about here? A code snippet demonstrating a change would be helpful in understanding the reason for this change.

Lenny4 commented:
Hi @coreymckrill ,

No problem :+1: 

I explained what my changes do here: https://drive.google.com/file/d/1XGsOv6YW0vPpsk7ku6PF8AH-IghlroTl/view?usp=sharing

Please tell me if it's clear

Lenny4 commented:
@coreymckrill up

Lenny4 commented:
@coreymckrill up

Lenny4 commented:
@coreymckrill up

Lenny4 commented:
Hi @roykho ,

Seems like corey is AFK can you take a look at this PL and tell me if it is clear ?

Thank you

coreymckrill commented:
@Lenny4 thanks for the video, it's super helpful for understanding the reason for the change. We discussed this in our bug scrub today. There is some concern that this would be a backward-incompatible change. If there is some third-party code that assumes the "old" way of binding, it could be doing something that would no longer work with this change. Here's a JS Fiddle with a very simplified example:

https://jsfiddle.net/40xfpvLe/4/

The "third party" code is unbinding the click event and preventing the callback. But if you switch to the "new" way of binding, the callback is no longer prevented.

Can you think of a further change or an alternative approach that would preserve back-compatibility?

Lenny4 commented:
@coreymckrill Indeed !

Wow didn't think of that.

Well in this case the only possible solution I see is to somehow make all functions called by the events, callable from a third party code.

I'm not sure if you guys want to implements it.

### How it would works ?

https://drive.google.com/file/d/1wj63flKBV8IjUJ626bH4tqY1iWHkQwn-/view?usp=sharing

So we would have to change the file `woocommerce/assets/js/admin/meta-boxes-order.js`

from:
```
wc_meta_boxes_order.init();
wc_meta_boxes_order_items.init();
wc_meta_boxes_order_notes.init();
wc_meta_boxes_order_downloads.init();
wc_meta_boxes_order_custom_meta.init();
```

to something like:
```log
window.globalWoocommerce = {
  wc_meta_boxes_order: wc_meta_boxes_order,
  wc_meta_boxes_order_items: wc_meta_boxes_order_items,
  wc_meta_boxes_order_notes: wc_meta_boxes_order_notes,
  wc_meta_boxes_order_downloads: wc_meta_boxes_order_downloads,
  wc_meta_boxes_order_custom_meta: wc_meta_boxes_order_custom_meta,
};
wc_meta_boxes_order.init();
wc_meta_boxes_order_items.init();
wc_meta_boxes_order_notes.init();
wc_meta_boxes_order_downloads.init();
wc_meta_boxes_order_custom_meta.init();
```
This way every third party plugin could call whatever function they want whenever they want to register again all the events

Please tell me what you think.

coreymckrill commented:
Interesting, thanks for thinking further about this. We'll be at a meetup next week and probably won't be able to respond until we get back but we'll discuss.

Lenny4 commented:
@coreymckrill up

Lenny4 commented:
@coreymckrill up

Lenny4 commented:
@coreymckrill did you have the meetup ?

coreymckrill commented:
@Lenny4 sorry again, we have had a lot going on since our meetup. We chatted a bit about your new suggested approach, and it sounds promising. Basically we are just exporting the various metabox objects into the global namespace so that other extensions can access their methods, such as `init`, yes? Do you think there is any downside or risk to having all the other methods accessible as well?

I think you should go ahead and update this PR to take that approach. Note that in that same script file we have another object that we're exporting, and we're naming it `window.wcTracks`. So following that convention, perhaps the best name for this would be something like `window.wcOrderMetaBoxes`?

Once the PR is updated, please provide some step-by-step testing instructions (with a code snippet if necessary) that demonstrate what this change does. There will be multiple people reviewing and testing this before it makes it into a release, and they won't all have time to watch the videos you made.

Lenny4 committed:
create global to call wc order meta boxe

Lenny4 committed:
Merge branch 'JQUERY_ON_DOCUMENT' of github.com:Lenny4/woocommerce into JQUERY_ON_DOCUMENT

Lenny4 commented:
Hi @coreymckrill ,

I updated the PR but I cannot think of a way to test this as woocommerce won't use `window.wcOrderMetaBoxes` but it's only there for third party plugin.

So I just added a commentary to say `this allows third party plugin to call woocommerce function`

What do you think ?

coreymckrill commented:
@Lenny4 here's what I meant by providing a code snippet:
https://gist.github.com/coreymckrill/a0ea623a0b98f1475144da04999947ae

If you add that to your test site in `wp-content/mu-plugins`, and then go to the Edit Order screen, you should see a new button at the top that says "Run Test". Clicking it will reload the contents of the Order Data meta box. Do you think this is sufficient to demonstrate and test the changes in this PR?

Lenny4 commented:
@coreymckrill 

Ok I anderstand now. Yes I think it would be sufficient.
I will try your code (and update it if needed) and provide the steps to reproduce.

Lenny4 commented:
Hi @coreymckrill 

Sorry for the late answer I was busy.

So I modified your code to add a 2nd button to show the difference.
Here is the code snippet:
https://gist.github.com/Lenny4/448cc3894b940d3877f02754a9893cdf

Do I have to do anything else ?

coreymckrill commented on plugins/woocommerce/changelog/51185-JQUERY_ON_DOCUMENT:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: tweak
+Comment: Not needed as it doesn't change how woocommerce is working or is used.
+
```
```suggestion
Significance: minor
Type: update

Make the scripts that initialize the meta boxes on the Edit Order screen accessible to third parties

```

coreymckrill replied:
```suggestion
Significance: minor
Type: enhancement

Make the scripts that initialize the meta boxes on the Edit Order screen accessible to third parties
```

😕 The CI changelog automation reverted the suggestion I made earlier. I made sure it won't do it again, just going to update this again.

coreymckrill reviewed (commented):
Hi @Lenny4 I updated the testing instructions on this, and also think the changelog should have details about this change, so I made a suggestion there. I'm going to ask for one more person to review before we merge this.

Lenny4 committed:
Update plugins/woocommerce/changelog/51185-JQUERY_ON_DOCUMENT

Co-authored-by: Corey McKrill <916023+coreymckrill@users.noreply.github.com>

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

coreymckrill committed:
Update plugins/woocommerce/changelog/51185-JQUERY_ON_DOCUMENT

barryhughes reviewed (approved):
Looks good to me, and tests as per the instructions :+1:

coreymckrill commented:
@Lenny4 Unfortunately we had to revert this because of some concerns that came up during testing before the release about exposing the order meta box objects to the global context. More info [here](https://github.com/woocommerce/woocommerce/pull/54390).

As an alternative approach, perhaps we could have all of the `init()` functions hooked to a custom event, that could then be triggered by external code when things need to be re-initialized?

Lenny4 commented:
Hi @coreymckrill 

Sorry for the slow answer

Something like this ?

```javascript
window.addEventListener('wc_meta_boxes_order_init', (e) => {
    wc_meta_boxes_order.init()
});

window.addEventListener('wc_meta_boxes_order_items_init', (e) => {
    wc_meta_boxes_order_items.init()
});

window.addEventListener('wc_meta_boxes_order_notes_init', (e) => {
    wc_meta_boxes_order_notes.init()
});

window.addEventListener('wc_meta_boxes_order_downloads_init', (e) => {
    wc_meta_boxes_order_downloads.init()
});

window.addEventListener('wc_meta_boxes_order_custom_meta_init', (e) => {
    wc_meta_boxes_order_custom_meta.init()
});
```
Can be trigger like this:

```javascript
window.dispatchEvent(new CustomEvent("wc_meta_boxes_order_init"));
```

Lenny4 commented:
Do I open another Pull request ?

coreymckrill commented:
@Lenny4 Yeah, that approach looks worth exploring. A new pull request would be the best way to do it.
